cmd/skopeo/seed.go

```diff
@@ -0,0 +1,91 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/containers/image/torrent"
+	"github.com/containers/image/transports/alltransports"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/urfave/cli"
+)
+
+func seedHandler(c *cli.Context) error {
+	policyContext, err := getPolicyContext(c)
+	if err != nil {
+		return fmt.Errorf("Error loading trust policy: %v", err)
+	}
+	defer policyContext.Destroy()
+
+	trackers := c.StringSlice("torrent-trackers")
+
+	ctx, err := contextFromGlobalOptions(c, "")
+	if err != nil {
+		return err
+	}
+	ctx.DockerTorrentTrackers = trackers
+
+	if len(c.Args()) < 2 || len(c.Args())%2 != 0 {
+		cli.ShowCommandHelp(c, "seed")
+		return errors.New("Invalid number of arguments")
+	}
+
+	debug := logrus.GetLevel() == logrus.DebugLevel
+
+	client, err := torrent.MakeClient(ctx, debug, true)
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	for off := 0; off < len(c.Args()); off = off + 2 {
+		ref, err := alltransports.ParseImageName(c.Args()[off])
+		if err != nil {
+			return fmt.Errorf("Invalid source image %s: %v", c.Args()[off], err)
+		}
+
+		refSrc, err := alltransports.ParseImageName(c.Args()[off+1])
+		if err != nil {
+			return fmt.Errorf("Invalid storage image %s: %v", c.Args()[off+1], err)
+		}
+
+		if err := client.Seed(context.Background(), ctx, ref, refSrc); err != nil {
```

Comment:
If I am reading the instructions right:
> ```console
> $ skopeo copy docker://quay.io/giuseppe/hello-world dir:/tmp/seed
> ```

This copies the image to a local directory.

> ```console
> $ skopeo seed --torrent-trackers udp4://192.168.125.222:6969
> ```

… and this makes _yet another_ copy of the blobs inside `torrent.Client.dir`. Is that really necessary to do both?

It does sound useful to seed an image from a local source for the benefit of a remote destination… maybe it’s just the instructions that could be updated to show that the `dir:` copy is optional; or am I misunderstanding this?

Reply:
no another copy is done.  It probably makes no sense to try to support other transports, and we can force `dir:` is used?  In this case I can try to tweak the torrent client to read directly from the `dir:` storage

Reply:
I’m not quite thrilled about code reaching into the internals of `dir:` (I assume the seed code requires random access, so `GetBlob` is insufficient); if this works just as well with

```console
dockerImageName=docker://quay.io/giuseppe/hello-world
skopeo seed --torrent-trackers udp4://192.168.125.222:6969 \
    $dockerImageName $dockerImageName
```
, with only the initial copy taking some time, then it’s just a matter of documentation for me.

Reply:
what I don't really like is how to specify the fully qualified image, as we need to access the tracker there or at least download the .torrent file (e.g. http://quay.io/c1/torrent/giuseppe/hello-world/blobs/sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4)

So we need two arguments for each image, the fully qualified name and the local storage.

Is there a better way for doing it?

Reply:
/cc @rhatdan

Reply:
\*shrug\* I’m fine with specifying two locations for source/destination (after all `skopeo copy` requires two locations as well, and this is conceptually fairly similar), I was really only speaking up against the _example_ causing two copies when one would do.